R/zchunk_LA118.hydro.R

```diff
@@ -29,41 +29,177 @@ module_energy_LA118.hydro_DISABLED <- function(command, ...) {
+
+    # Calculation of economic hydropower potential by country, in EJ/yr
+    # Calculate a capacity factor for translating MW to GWh, using weighted average capacity factor of all existing dams
+    Hydropower_potential %>%
+      gather(variable, value, -Country, -iso) %>%
+      group_by(variable) %>%
+      summarise(value = sum(value, na.rm = TRUE)) %>%
+      # Calculation of capacity factor
+      summarise(value = value[variable == "Installed_GWh"] / (value[variable == "Installed_MW"] * CONV_YEAR_HOURS * CONV_MIL_BIL)) %>%
+      .$value -> # Convert table to single number
+      Hydro_capfac
+
+    # Economic potential is what we are interested in from this database; however it is often not reported. Many countries without reported
+    # economic potential nevertheless have estimates of the technical potential.
+    # Calculate a translation from Technical potential to Economic potential, using weighted average among regions where both are reported
+    # First, for countries reporting in MW, convert to GWh (most countries have potentials as GWh but some are in MW)
+    Hydropower_potential %>%
+      mutate_if(is.integer, as.numeric) %>% # Convert columns that are getting read in as integers to numbers
+      mutate(Technical_GWh = if_else(is.na(Technical_GWh), Technical_MW * CONV_YEAR_HOURS * CONV_MIL_BIL * Hydro_capfac, Technical_GWh)) %>%
+      mutate(Economic_GWh = if_else(is.na(Economic_GWh), Economic_MW * CONV_YEAR_HOURS * CONV_MIL_BIL * Hydro_capfac, Economic_GWh)) ->
+      Hydropower_potential
+
+    # Among countries with both technical and economic potential reported, calculate an average translation from one to the other
+    Hydropower_potential %>%
+      summarise(value = sum(Economic_GWh[!is.na(Technical_GWh) & !is.na(Economic_GWh)]) / sum(Technical_GWh[!is.na(Technical_GWh) & !is.na(Economic_GWh)])) %>%
+      .$value -> # Convert table to single number
+      Hydro_tech_econ
+
+    # For countries with technical potential reported but no economic potential, estimate the economic potential
+    Hydropower_potential %>%
+      mutate(Economic_GWh = if_else(is.na(Economic_GWh), Technical_GWh * Hydro_tech_econ, Economic_GWh)) %>%
+      # This still leaves a few countries for which no estimates of hydro potentials are provided.
+      # The largest among them is North Korea; most of the others are islands or deserts that probably don't have much beyond present-day production.
+      # Not worrying about these (future growth rates will be set to 0)
+      mutate(Economic_EJ = Economic_GWh * CONV_GWH_EJ) %>% # Convert GWh to EJ
+      select(iso, Economic_EJ) -> # Stripping down unneeded information from table
+      Hydropower_potential
+
+    # Calculate the growth potential by country, as the economic potential minus the actual generation in the most recent historical year (from the IEA balances)
+      L100.IEA_en_bal_ctry_hist %>%
+        filter(FLOW == "ELOUTPUT", PRODUCT == "Hydro", year == 2010) %>%
```

Comment:
2010 should be a defined constant, or `max(...)` something, like HISTORICAL_YEARS